src/host-config/proxy.ts

```diff
@@ -0,0 +1,275 @@
+import { promises as fs } from 'fs';
+import path from 'path';
+import { isRight } from 'fp-ts/lib/Either';
+import Reporter from 'io-ts-reporters';
+
+import type { RedsocksConfig, HostProxyConfig } from './types';
+import { ProxyConfig } from './types';
+import { pathOnBoot, readFromBoot, writeToBoot } from '../lib/host-utils';
+import { unlinkAll, mkdirp } from '../lib/fs-utils';
+import { isENOENT } from '../lib/errors';
+import log from '../lib/supervisor-console';
+import * as dbus from '../lib/dbus';
+
+const proxyBasePath = pathOnBoot('system-proxy');
+const noProxyPath = path.join(proxyBasePath, 'no_proxy');
+const redsocksConfPath = path.join(proxyBasePath, 'redsocks.conf');
+
+export const disallowedProxyFields = ['local_ip', 'local_port'];
+export const validRedsocksProxyTypes = [
+	'socks4',
+	'socks5',
+	'http-connect',
+	'http-relay',
+];
+
+const isAllowedProxyField = (field: string): boolean =>
+	['type', 'ip', 'port', 'login', 'password'].includes(field);
+
+const isAuthField = (field: string): boolean =>
+	['login', 'password'].includes(field);
+
+// ? is a lazy operator, so only the contents up until the first `}(?=\s|$)` is matched.
+// (?=\s|$) indicates that `}` must be followed by a whitespace or end of file to match,
+// in case there are user fields with brackets such as login or password fields.
+const blockRegexFor = (blockLabel: string) =>
+	new RegExp(`${blockLabel}\\s?{([\\s\\S]+?)}(?=\\s|$)`);
+
+const baseBlock = {
+	log_debug: 'off',
+	log_info: 'on',
+	log: 'stderr',
+	daemon: 'off',
+	redirector: 'iptables',
+};
+
+export class RedsocksConf {
+	public static stringify(config: RedsocksConfig): string {
+		const blocks: string[] = [];
+
+		if (config.redsocks && Object.keys(config.redsocks).length > 0) {
+			blocks.push(RedsocksConf.stringifyBlock('base', baseBlock));
+			blocks.push(
+				RedsocksConf.stringifyBlock('redsocks', {
+					...config.redsocks,
+					local_ip: '127.0.0.1',
+					local_port: 12345,
+				}),
+			);
+		}
+
+		return blocks.length ? blocks.join('\n') : '';
+	}
+
+	public static parse(rawConf: string): RedsocksConfig {
+		const conf: RedsocksConfig = {};
```

Comment:
Is good practice to trim strings in case they are just whitespace

```suggestion
		const conf: RedsocksConfig = {};
		rawConf = rawConf.trim();
```